pallets/hyperdrive/src/tezos.rs

```diff
@@ -0,0 +1,570 @@
+where
+    ParsableAccountId: FromStr + Into<AccountId>,
+    Extra: From<RegistrationExtra<Reward, Balance, AccountId>>,
+    Reward: Parameter + Member + TryFrom<Vec<u8>>,
+    Balance: From<u128>,
+{
+    type Error = ValidationError;
+    fn parse(encoded: &[u8]) -> Result<ParsedAction<AccountId, Extra>, ValidationError> {
+        let (action, origin, payload) = parse_message(encoded)?;
+
+        Ok(match action {
+            RawAction::RegisterJob => {
+                let payload: Vec<u8> = (&payload).into();
+                let (job_id_sequence, registration) = parse_job_registration_payload::<
+                    Reward,
+                    Balance,
+                    ParsableAccountId,
+                    AccountId,
+                    Extra,
+                >(payload.as_slice())?;
+
+                ParsedAction::RegisterJob(
+                    (to_multi_origin(&origin)?, job_id_sequence),
+                    registration,
+                )
+            }
+        })
+    }
+}
+
+impl<T: Config, I> From<ValidationError> for Error<T, I> {
+    fn from(_: ValidationError) -> Self {
+        Error::<T, I>::MessageParsingFailed
+    }
+}
+
+#[cfg_attr(rustfmt, rustfmt::skip)]
+fn message_schema() -> &'static Micheline {
+    static MESSAGE_SCHEMA: OnceBox<Micheline> = OnceBox::new();
+    MESSAGE_SCHEMA.get_or_init(|| {
+        let schema: Micheline = pair(vec![
+            string(),
+            address(),
+            bytes(),
+        ]);
+        Box::new(schema)
+    })
+}
+
+/// Parses an encoded message from Tezos representing an action into a tuple `(ACTION, ORIGIN, PAYLOAD)`.
+///
+/// # Example
+/// A message to register a job could look like:
+///
+/// ```txt
+/// PrimitiveApplication(PrimitiveApplication { prim: "Pair", args: Some([Literal(String(String("REGISTER_JOB"))), PrimitiveApplication(PrimitiveApplication { prim: "Pair", args: Some([Literal(Bytes(Bytes("0x00008a8584be3718453e78923713a6966202b05f99c6"))), Literal(Bytes(Bytes("0x050707030a0707030607070a0000001601000000000000000000000000000000000000000000070707070000070703060707030607070a00000035697066733a2f2f516d64484c6942596174626e6150645573544d4d4746574534326353414a43485937426f3741445832636444656100010707000207070001070700010707070700b0d403070700bfe6d987d86107070098e4030707000000bf9a9f87d86107070a00000035697066733a2f2f516d64484c6942596174626e6150645573544d4d4746574534326353414a43485937426f374144583263644465610001")))]), annots: None })]), annots: None })
+/// ```
+fn parse_message(encoded: &[u8]) -> Result<(RawAction, TezosAddress, Bytes), ValidationError> {
+    let unpacked: Micheline = Micheline::unpack(encoded, Some(message_schema()))
+        .map_err(|e| ValidationError::TezosMicheline(e))?;
+
+    let p: PrimitiveApplication = unpacked.try_into()?;
+    let pair: Pair = p.try_into()?;
+
+    let values = pair.flatten().values;
+    if values.len() != 3 {
+        Err(ValidationError::InvalidMessage)?;
+    }
```

Comment:
This is already checked by the unpacking.
```suggestion
```

Reply:
it probably does now, but will it do in all future?

Reply:
Yes, the bytes need to match the type.